src/index.adoc

```diff
@@ -26,225 +45,301 @@ shown here.
+| Message length | 64-bit <<uLEB>> | The length of the message in bytes
+| Message | UTF-8 encoded string | The message associated with this change
+| Other actors | <<Array of Actor IDs>> | Other actor IDs in this change
+| Operation columns metadata | <<Column Metadata>> | Description of the <<Operation Columns,operation columns>>
+| Operation columns | <<Column Data>> | The actual bytes for the operation columns
+| Extra bytes | bytes | All data remaining in the chunk
+|===
+
+A change chunk just contains a single change, its metadata and operations. It
+does not include any dependent changes, so you can only apply the change to a
+document that already contains those dependent changes.
+
+Change chunks use a columnar storage format that assumes that the values of
+various fields are similar across adjacent operations to optimize for high
+compression ratios and fast decoding.
+
+The extra bytes must be retained when processing changes. If future versions of
+automerge add new metadata to changes, this will allow old clients to
+collaborate with new clients without limiting which features the new clients can
+use.
+
+[#compressed-chunks]
+==== Compressed Chunk
+
+Compressed chunks must be decompressed using <<DEFLATE>>. The decompressed chunk
+should be either a <<Document Chunk>> or <<Change Chunk>>. Implementations
+SHOULD raise an error if the contents of a compressed chunk is another
+compressed chunk.
 
 === Simple types
 
-==== uLEB and LEB
+==== uLEB
 
 uLEB is an unsigned https://en.wikipedia.org/wiki/LEB128[little endian base 128] value.
-This is a variable length encoding used throughout this document.
+This is a variable length encoding used throughout.
 
-LEB is the signed variant.
+To encode a uLEB, represent the number in binary and pad it with leading zeros
+so that it has a length which is a multiple of 7. Take each group of 7 bytes from
+least-significant to most-significant and output them in bytes - the first bit
+of every byte is 1 except for the last byte which is 0.
 
-[#action-array]
-==== Action array
+* Unsigned ints 0 - 127 are stored as one byte: `0b00000000 - 0b01111111`
+* Unsigned ints 128 - 16383 are stored as two bytes: `0b10000000 0b00000001 - 0b11111111 0b01111111`
+etc.
 
-The actions of the reference data model are encoded in the storage format as a
-0-based index into the following array:
+To decode a uLEB, read bytes up to and including the first byte with a 0 as the
+first bit.  Take the latter 7 bits from each byte (the last byte contains the
+most significant bits, so you need to concatenate them in the opposite order to
+which the bytes are represented on disk).
 
-|===
-| Action
-
-| `makeMap`
-| `set`
-| `makeList`
-| `del`
-| `makeText`
-| `inc`
-| `makeTable`
-| `link`
-|===
+Although uLEB encoding can support numbers of arbitrary bitsize, fields in
+Automerge are size limited to 32 or 64 bits. Implementations should fail to
+parse documents that contain overly large encoded integers in fields.
```

Comment:
:+1: Added (and same to the LEB section).

For future changes: I think the only place we use a 32-bit LEB currently is the column specification. We could describe that by limiting the size of the column id to 2**28 instead of having two differently sized options.